### ISSUE TITLE
Stream error is thrown when InputStream is read

### ISSUE THREAD

**Issue by alissapajer** [typelevel/fs2#2079]
```scala
scala> implicit val cs: ContextShift[IO] = IO.contextShift(scala.concurrent.ExecutionContext.Implicits.global)
cs: cats.effect.ContextShift[cats.effect.IO] = cats.effect.internals.IOContextShift@ff60478

scala> val blocker = Blocker.liftExecutionContext(scala.concurrent.ExecutionContext.Implicits.global)
blocker: cats.effect.Blocker = cats.effect.Blocker@3eebbccb

scala> Stream.raiseError[IO](new Throwable("no way")).through(io.toInputStream).flatMap(is => Stream.eval(blocker.delay[IO, Int](is.read(Array[Byte]()))))
res1: fs2.Stream[[x]cats.effect.IO[x],slamdata.Predef.Int] = Stream(..)

scala> res1.compile.toList.attempt.unsafeRunSync()
java.lang.Throwable: no way
        at $line8.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$.<init>(<console>:35)
        at $line8.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$.<clinit>(<console>)
        at $line8.$eval$.$print$lzycompute(<console>:7)
        at $line8.$eval$.$print(<console>:6)
        at $line8.$eval.$print(<console>)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.call(IMain.scala:745)
        at scala.tools.nsc.interpreter.IMain$Request.loadAndRun(IMain.scala:1021)
        at scala.tools.nsc.interpreter.IMain.$anonfun$interpret$1(IMain.scala:574)
        at scala.reflect.internal.util.ScalaClassLoader.asContext(ScalaClassLoader.scala:41)
        at scala.reflect.internal.util.ScalaClassLoader.asContext$(ScalaClassLoader.scala:37)
        at scala.reflect.internal.util.AbstractFileClassLoader.asContext(AbstractFileClassLoader.scala:41)
        at scala.tools.nsc.interpreter.IMain.loadAndRunReq$1(IMain.scala:573)
        at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:600)
        at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:570)
        at scala.tools.nsc.interpreter.ILoop.interpretStartingWith(ILoop.scala:894)
        at scala.tools.nsc.interpreter.ILoop.command(ILoop.scala:762)
        at scala.tools.nsc.interpreter.ILoop.processLine(ILoop.scala:464)
        at scala.tools.nsc.interpreter.ILoop.loop(ILoop.scala:485)
        at scala.tools.nsc.interpreter.ILoop.process(ILoop.scala:1077)
        at xsbt.ConsoleInterface.run(ConsoleInterface.scala:75)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at sbt.internal.inc.AnalyzingCompiler.call(AnalyzingCompiler.scala:248)
        at sbt.internal.inc.AnalyzingCompiler.console(AnalyzingCompiler.scala:210)
        at sbt.Console.console0$1(Console.scala:48)
        at sbt.Console.$anonfun$apply$2(Console.scala:51)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at sbt.util.InterfaceUtil$$anon$1.get(InterfaceUtil.scala:10)
        at sbt.TrapExit$App.run(TrapExit.scala:257)
        at java.lang.Thread.run(Thread.java:748)
res2: Either[Throwable,List[slamdata.Predef.Int]] = Left(java.io.IOException: UpStream failed)
```

I would expect the `Throwable` to not be thrown, which is what happens if we don't call `is.read`:
```scala
scala> Stream.raiseError[IO](new Throwable("no way")).through(io.toInputStream).flatMap(is => Stream.eval(blocker.delay[IO, InputStream](is)))
res3: fs2.Stream[[x]cats.effect.IO[x],slamdata.Predef.InputStream] = Stream(..)

scala> res3.compile.toList.attempt.unsafeRunSync()
res4: Either[Throwable,List[slamdata.Predef.InputStream]] = Left(java.lang.Throwable: no way)
```

Am I misusing the `Blocker` in any way?

**Comment by mpilquist** [comment]
Looks like the exception isn't getting thrown to the caller but rather reported to `ExecutionContext.reportFailure`, which is calling `printStackTrace`. Not sure why though -- it also works as expected in 3.x.

**Comment by mpilquist** [comment]
Okay, here's what's happening -- the `InputStream` is wrapped in a `Resource`. The call to `is.read` fails with the `UpStream failed` exception, which causes the overall stream to fail. The finalizer for the resource is run, which notices that upstream failed due to an exception and it also then fails. The CE2 `IO` type logs an exception thrown from a finalizer if the finalizer was being run due to a previous exception.

**Comment by alissapajer** [comment]
Ah right I see, it's returning `Left(java.io.IOException: UpStream failed)` and also printing the stack trace.

What I pasted above is a minimization of a test I wrote that actually was throwing, but it was throwing because I wasn't handling the correct error type in `ApplicativeError#recoverWith`. It was erroring with 
```
fs2.CompositeFailure: Multiple exceptions were thrown (3), first java.io.IOException: UpStream failed
```
but I was expecting the exception that actually caused the error.

So, since it's doing the right thing in 3.x, I'd be 👍 to close this issue. The stack trace printing threw me.
